### PR TITLE
Add MAE

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

This PR implements [MAE](https://github.com/facebookresearch/mae) (Masked Autoencoders are Scalable Vision Learners) by Facebook AI.

I've implemented it as a separate model in the library, as the model is quite specific (encoder only operates on visual patches, next a mask token is added and the decoder reconstructs pixel values based on encoded visual patches + mask tokens). The model without decoder on top is called `ViTMAEModel`, the model with decoder on top is called `ViTMAEForPreTraining` (inspired by `BertForPreTraining`). 

After pre-training, one can load the weights directly into a `ViTForImageClassification`.

* Models are on the hub: https://huggingface.co/models?other=vit_mae

* Demo notebook: https://colab.research.google.com/drive/1edgG0ne4VNQrc11uAS0wzogg5HblzK2s?usp=sharing

To do:

- [x] fix some tests. However, these fail due to the non-deterministic behaviour of the model (a random mask is generated in each forward pass). I can make the integration test deterministic by adding `torch.manual_seed(2)`, however adding this on top of the test file doesn't make it deterministic for other tests.

### CODE REVIEW

File: docs/source/model_doc/vit_mae.mdx [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+specific language governing permissions and limitations under the License.
+-->
+
+# ViTMAE
+
+## Overview
+
+The ViTMAE model was proposed in [Masked Autoencoders Are Scalable Vision Learners] by Kaiming He, Xinlei Chen, Saining Xie, Yanghao Li, Piotr Dollár, Ross Girshick.
```
Comment (sgugger) [inline docs/source/model_doc/vit_mae.mdx:17]: The link to the paper is missing.

File: src/transformers/models/vit_mae/configuration_vit_mae.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+            The dropout ratio for the attention probabilities.
+        initializer_range (`float`, *optional*, defaults to 0.02):
+            The standard deviation of the truncated_normal_initializer for initializing all weight matrices.
+        layer_norm_eps (`float`, *optional*, defaults to 1e-12):
+            The epsilon used by the layer normalization layers.
+        image_size (`int`, *optional*, defaults to `224`):
+            The size (resolution) of each image.
+        patch_size (`int`, *optional*, defaults to `16`):
+            The size (resolution) of each patch.
+        num_channels (`int`, *optional*, defaults to `3`):
```
Comment (sgugger) [inline src/transformers/models/vit_mae/configuration_vit_mae.py:64]:
```suggestion
        image_size (`int`, *optional*, defaults to 224):
            The size (resolution) of each image.
        patch_size (`int`, *optional*, defaults to 16):
            The size (resolution) of each patch.
        num_channels (`int`, *optional*, defaults to 3):
```
We don't put numbers in code.

File: src/transformers/models/vit_mae/configuration_vit_mae.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+            Whether to add a bias to the queries, keys and values.
+        decoder_num_attention_heads (`int`, *optional*, defaults to `12`):
+            Number of attention heads for each attention layer in the decoder.
+        decoder_hidden_size (`int`, *optional*, defaults to `512`):
+            Dimensionality of the decoder.
+        decoder_num_hidden_layers (`int`, *optional*, defaults to `8`):
+            Number of hidden layers in the decoder.
+        decoder_intermediate_size (`int`, *optional*, defaults to `2048`):
+            Dimensionality of the "intermediate" (i.e., feed-forward) layer in the decoder.
+        mask_ratio (`float`, *optional*, defaults to `0.75`):
```
Comment (sgugger) [inline src/transformers/models/vit_mae/configuration_vit_mae.py:76]: Same here.

File: src/transformers/models/vit_mae/feature_extraction_vit_mae.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    ImageInput,
+    is_torch_tensor,
+)
+from ...utils import logging
+
+
+logger = logging.get_logger(__name__)
+
+
+class ViTMAEFeatureExtractor(FeatureExtractionMixin, ImageFeatureExtractionMixin):
```
Comment (sgugger) [inline src/transformers/models/vit_mae/feature_extraction_vit_mae.py:37]: It looks exactly like `ViTFeatureExtractor`. If there are no differences, the `ViTFeatureExtractor` should be reused instead of creating a new class (and you can map `ViTMAEConfig` to `ViTFeatureExtractor` in the auto mapping).

File: src/transformers/models/vit_mae/inference.py [hunk 5]
```diff
@@ -0,0 +1,13 @@
+import torch
```
Comment (sgugger) [inline src/transformers/models/vit_mae/inference.py:1]: This file should be removed.

File: src/transformers/models/vit_mae/modeling_vit_mae.py [hunk 6]
```diff
@@ -0,0 +1,941 @@
...
+    """
+    Class for ViTMAEModel's outputs, with potential hidden states and attentions.
+
+    Args:
+        last_hidden_state (`torch.FloatTensor` of shape `(batch_size, sequence_length, hidden_size)`):
+            Sequence of hidden-states at the output of the last layer of the model.
+        mask
+            ...
+        ids_restore
+            ...
```
Comment (sgugger) [inline src/transformers/models/vit_mae/modeling_vit_mae.py:64]: This needs to be filled.

File: src/transformers/models/vit_mae/modeling_vit_mae.py [hunk 7]
```diff
@@ -0,0 +1,941 @@
...
+
+    Args:
+        loss (`torch.FloatTensor` of shape `(1,)`):
+            Pixel reconstruction loss.
+        logits (`torch.FloatTensor` of shape `(batch_size, patch_size ** 2 * num_channels)`):
+            Pixel reconstruction logits.
+        mask
+            ...
+        ids_restore
+            ...
```
Comment (sgugger) [inline src/transformers/models/vit_mae/modeling_vit_mae.py:118]: This needs to be filled.

File: src/transformers/models/vit_mae/modeling_vit_mae.py [hunk 8]
```diff
@@ -0,0 +1,941 @@
...
+
+    grid = grid.reshape([2, 1, grid_size, grid_size])
+    pos_embed = get_2d_sincos_pos_embed_from_grid(embed_dim, grid)
+    if cls_token:
+        pos_embed = np.concatenate([np.zeros([1, embed_dim]), pos_embed], axis=0)
+    return pos_embed
+
+
+def get_2d_sincos_pos_embed_from_grid(embed_dim, grid):
+    assert embed_dim % 2 == 0
```
Comment (sgugger) [inline src/transformers/models/vit_mae/modeling_vit_mae.py:164]: No assert in modeling files please.

File: src/transformers/models/vit_mae/modeling_vit_mae.py [hunk 9]
```diff
@@ -0,0 +1,941 @@
...
+
+    emb = np.concatenate([emb_h, emb_w], axis=1)  # (H*W, D)
+    return emb
+
+
+def get_1d_sincos_pos_embed_from_grid(embed_dim, pos):
+    """
+    embed_dim: output dimension for each position pos: a list of positions to be encoded: size (M,) out: (M, D)
+    """
+    assert embed_dim % 2 == 0
```
Comment (sgugger) [inline src/transformers/models/vit_mae/modeling_vit_mae.py:178]: Same here.

File: src/transformers/models/vit_mae/modeling_vit_mae.py [hunk 10]
```diff
@@ -0,0 +1,941 @@
...
+# From PyTorch internals
+def to_2tuple(x):
+    if isinstance(x, collections.abc.Iterable):
+        return x
+    return (x, x)
+
+
+def get_2d_sincos_pos_embed(embed_dim, grid_size, cls_token=False):
+    """
+    grid_size: int of the grid height and width return: pos_embed: [grid_size*grid_size, embed_dim] or
```
Comment (sgugger) [inline src/transformers/models/vit_mae/modeling_vit_mae.py:148]: If we have a docstring, let's describe what the function does and use the regular syntax for docstrings.

File: src/transformers/models/vit_mae/modeling_vit_mae.py [hunk 11]
```diff
@@ -0,0 +1,941 @@
...
+# Inspired by
+# https://github.com/rwightman/pytorch-image-models/blob/b9bd960a032c75ca6b808ddeed76bee5f3ed4972/timm/models/layers/helpers.py
+# From PyTorch internals
+def to_2tuple(x):
+    if isinstance(x, collections.abc.Iterable):
+        return x
+    return (x, x)
+
+
+def get_2d_sincos_pos_embed(embed_dim, grid_size, cls_token=False):
```
Comment (sgugger) [inline src/transformers/models/vit_mae/modeling_vit_mae.py:146]: `cls_token` is not a good name as I would expect a token, not a bool. `add_cls_token` seems better.

File: src/transformers/models/vit_mae/modeling_vit_mae.py [hunk 12]
```diff
@@ -0,0 +1,941 @@
...
+
+        # timm's trunc_normal_(std=.02) is effectively normal_(std=0.02) as cutoff is too big (2.)
+        torch.nn.init.normal_(self.cls_token, std=0.02)
+
+    def random_masking(self, x):
+        """
+        Perform per-sample random masking by per-sample shuffling. Per-sample shuffling is done by argsort random
+        noise. x: [N, L, D], sequence
+        """
+        N, L, D = x.shape  # batch, length, dim
```
Comment (sgugger) [inline src/transformers/models/vit_mae/modeling_vit_mae.py:237]: If those are `batch, length, dim` then let's use those names instead of setting a comment.

Also those are not classes, so should not be capitalized and we really shouldn't use one-letter variable names.

File: src/transformers/models/vit_mae/modeling_vit_mae.py [hunk 13]
```diff
@@ -0,0 +1,941 @@
...
+
+        ```python
+        >>> from transformers import ViTMAEFeatureExtractor, ViTMAEModel
+        >>> from PIL import Image
+        >>> import requests
+
+        >>> url = "http://images.cocodataset.org/val2017/000000039769.jpg"
+        >>> image = Image.open(requests.get(url, stream=True).raw)
+
+        >>> feature_extractor = ViTMAEFeatureExtractor.from_pretrained("facebook/vit-mae-base")
```
Comment (sgugger) [inline src/transformers/models/vit_mae/modeling_vit_mae.py:904]: Use the `ViTFeatureExtractor` here if the class is removed.

File: tests/test_modeling_vit_mae.py [hunk 14]
```diff
@@ -0,0 +1,374 @@
...
+    """
+
+    all_model_classes = (
+        (
+            ViTMAEModel,
+            ViTMAEForPreTraining,
+        )
+        if is_torch_available()
+        else ()
+    )
```
Comment (sgugger) [inline tests/test_modeling_vit_mae.py:162]: All fits in one line.

File: utils/check_repo.py [hunk 15]
```diff
@@ -103,6 +103,7 @@
 # should **not** be the rule.
 IGNORE_NON_AUTO_CONFIGURED = PRIVATE_MODELS.copy() + [
     # models to ignore for model xxx mapping
+    "ViTMAEForPreTraining",
```
Comment (sgugger) [inline utils/check_repo.py:106]: Why is this not added to the `ForPreTraining` auto-mapping? This one does not have a unified API so every model should be in it.

**Review by sgugger — APPROVED** [review]
Thanks for adding this new model!

File: src/transformers/models/vit_mae/feature_extraction_vit_mae.py (hunk 4 above)
Comment (NielsRogge) [reply src/transformers/models/vit_mae/feature_extraction_vit_mae.py:37]: It is identical, except that it uses the ImageNet default mean and std instead of the standard ones.

Will this still work? Should I just upload the `preprocessor_config.json` to the models on the hub, to make sure they use the appropriate mean and std?

File: utils/check_repo.py (hunk 15 above)
Comment (NielsRogge) [reply utils/check_repo.py:106]: Ok, I'll add it.

File: src/transformers/models/vit_mae/feature_extraction_vit_mae.py (hunk 4 above)
Comment (sgugger) [reply src/transformers/models/vit_mae/feature_extraction_vit_mae.py:37]: Yes, with the right preprocessor config, it should work.

**Review by LysandreJik — APPROVED** [review]
Thank you for your work, @NielsRogge!
